infra/events/src/main/java/com/simprints/infra/events/event/local/EventLocalDataSource.kt

```diff
@@ -1,25 +1,151 @@
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
 
-internal interface EventLocalDataSource {
+internal open class EventLocalDataSource @Inject constructor(
+    private val eventDatabaseFactory: EventDatabaseFactory,
+    private val jsonHelper: JsonHelper,
+    @DispatcherIO private val readingDispatcher: CoroutineDispatcher,
+    @NonCancellableIO private val writingContext: CoroutineContext,
+) {
 
-    suspend fun saveSessionScope(scope: SessionScope)
-    suspend fun countSessions(): Int
-    suspend fun loadOpenedSessions(): List<SessionScope>
-    suspend fun loadClosedSessions(): List<SessionScope>
-    suspend fun deleteSession(sessionId: String)
+    private var eventDao: EventRoomDao = eventDatabaseFactory.build().eventDao
 
-    suspend fun saveEvent(event: Event)
-    suspend fun observeEventCount(): Flow<Int>
-    suspend fun observeEventCount(type: EventType): Flow<Int>
-    suspend fun loadAllEvents(): Flow<Event>
-    suspend fun loadEventJsonInSession(sessionId: String): List<String>
-    suspend fun loadEventsInSession(sessionId: String): List<Event>
-    suspend fun deleteEventsInSession(sessionId: String)
+    private var scopeDao: SessionScopeRoomDao = eventDatabaseFactory.build().scopeDao
 
-    suspend fun deleteAll()
+    private val mutex = Mutex()
+
+    private suspend fun <R> useRoom(context: CoroutineContext, block: suspend () -> R): R =
+        withContext(context) {
+            try {
+                block()
+            } catch (ex: SQLiteException) {
+                if (isFileCorruption(ex)) {
+                    rebuildDatabase(ex)
+                    // Retry operation with new file and key
+                    block()
+                } else {
+                    throw ex
+                }
+            }
+        }
+
+    private suspend fun <R> useRoomFlow(context: CoroutineContext, block: () -> Flow<R>): Flow<R> =
+        withContext(context) {
+            try {
+                block().catch { cause ->
+                    if (isFileCorruption(cause)) {
+                        rebuildDatabase(cause)
+                        // Recreate flow and re-emit values with the new file and key
+                        emitAll(block())
+                    } else {
+                        throw cause
+                    }
+                }
+            } catch (ex: SQLiteException) {
+                if (isFileCorruption(ex)) {
+                    rebuildDatabase(ex)
+                    // Recreate flow with the new file and key
+                    block()
+                } else {
+                    throw ex
+                }
+            }
+        }
+
+    private fun isFileCorruption(ex: Throwable) =
+        ex is SQLiteDatabaseCorruptException ||
+            ex.let { it as? SQLiteException }?.message?.contains("file is not a database") == true
```

Comment:
We might want to add `ignoreCase=true` here just in case

Reply:
Same as the database factory -  this file shows up in diff because I removed the redundant internal interface and this code was previously in "Impl" file.